pkg/controller/namespace/namespace_controller.go

```diff
@@ -0,0 +1,165 @@
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a Namespace object and makes changes based on the state read
+// and what is in the Namespace.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  This example creates
+// a Pod as an example
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileNamespace) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling Namespace")
+
+	// Fetch the Namespace instance
+	instance := &corev1.Namespace{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	namespaceList := &corev1.NamespaceList{}
+	opts := client.ListOptions{Namespace: request.Namespace}
+	err = r.client.List(context.TODO(), &opts, namespaceList)
+	if err != nil {
+		reqLogger.Error(err, "Failed to get namespaceList")
+		return reconcile.Result{}, err
+	}
+
+	subjectPermissionList := &managedv1alpha1.SubjectPermissionList{}
+	opts = client.ListOptions{Namespace: request.Namespace}
+	err = r.client.List(context.TODO(), &opts, subjectPermissionList)
+	if err != nil {
+		reqLogger.Error(err, "Failed to get clusterRoleBindingList")
+		return reconcile.Result{}, err
+	}
+
+	clusterRoleList := &v1.ClusterRoleList{}
+	opts = client.ListOptions{Namespace: request.Namespace}
+	err = r.client.List(context.TODO(), &opts, clusterRoleList)
+	if err != nil {
+		reqLogger.Error(err, "Failed to get clusterRoleList")
+		return reconcile.Result{}, err
+	}
+
+	// loop through all subject permissions
+	for _, subjectPermission := range subjectPermissionList.Items {
+		// loop through all permissions in each
+		for _, permission := range subjectPermission.Spec.Permissions {
+			// 1st pass, appy allow regex
+			sl := controllerutil.AllowedNamespacesList(permission.NamespacesAllowedRegex, namespaceList)
+
+			// 2nd pass, apply deny regex
+			safeList := controllerutil.SafeListAfterDeniedRegex(permission.NamespacesDeniedRegex, sl)
+
+			// if namespace is in safeList, create RoleBinding
+			if namespaceInSlice(instance.Name, safeList) {
+				roleBinding := controllerutil.NewRoleBinding(permission.ClusterRoleName, subjectPermission.Spec.SubjectName, subjectPermission.Spec.SubjectKind, instance.Name)
```

Comment:
I have a general question about a potential race condition - (maybe @cblecker, @jharrington22, @rogbas might know(?))

What happens if someone is doing a `kubectl apply -f deploy/` and the first manifest in the directory is one which creates a `Namespace`. That will trigger a Reconcile and we get to this point in the code and then attempt to create a `RoleBinding`. At the same time, `kubectl` is progressing through the files in `deploy/` and gets to a manfiest that is creating a `RoleBinding` with the same name. Whether or not the manifest should do that or not, what is the behaviour here?

Is it possible that this line of code will attempt to create the `RoleBinding` that was just created by `kubectl`? Do we need to check for that in line 130 and treat it as a non-fatal error (or perhaps reconcile it in a different place)?

Reply:
It's possible, especially because this is a caching client. There's a couple ways to handle this:
- error and restart the loop, and let the next iteration of the loop handle it (this would need an if exists check prior to the create)
- gracefully handle the error and continue. the errors package provides methods to handle this: https://godoc.org/k8s.io/apimachinery/pkg/api/errors

You may also want to decide if you want to apply an ownersRef to an existing object if it already exists, IF you want to steal ownership of that object.

Reply:
Yeah either of those choices would be good. Essentially we just need extra protections, but should be fairly easy to just catch an error and handle it.